### DISCUSSION TITLE
Can I have multiple instances of Shlink with an external PostgreSQL DB, without a Redis server? How?

### DISCUSSION THREAD

**Discussion by PrateeksGitHub** [shlinkio/discussions#148] [Help wanted] [answered]
Hey all,
I want to run multiple instances of Shlink in my k8s cluster. I want to use my existing PostgreSQL DB and avoid using Redis (mainly due to cost considerations). I have a few questions:

- Is Redis mandatory for multiple instances of Shlink?
- What, if any, are some alternative ways?
- Is it required for the service to run or just during boot up?

I've referenced [this](https://shlink.io/documentation/install-docker-image/#use-an-external-redis-server) doc and docs it links to.

**Comment by acelaya** [comment, marked as answer]
I think those docs may require some re-wording.

Redis is not mandatory if you are running independent instances, with independent databases, as long as every instance has a single replica.

Redis is required for multi-replica instances (what is called there multi-instance), and there are no supported alternatives at the moment. You can see the considerations here https://shlink.io/documentation/install-docker-image/#multi-instance-considerations

**Reply by PrateeksGitHub** [reply]
Thanks @acelaya . I need use multi-replica instances.

Just need to clarify if Redis is required for anything other than "creating the database for the first time, or downloading the GeoLite2 database"? These are just one-time requirements, so can I get away by removing Redis after DB creation?

**Reply by acelaya** [reply]
No, it's not just about initial database creation, it is also used at least for database migrations.

Every time you update to new Shlink versions, there might be database migrations to be run. Without a shared redis, all replicas will try to run migrations at the same time, which could end up in failing containers, and even a corrupted database in the worst case scenario.

Migrations will also be run every time your services scales-up to more replicas, but in that case it might be less critical because it should result in no migrations being run.

In any case these are implementation details, and the plan is to be able to use this mechanism for other purposes. You can proceed without redis, and it might end up working now, but break in a future version. In that case you won't get official support because you have decided not to follow the documented path.

**Reply by PrateeksGitHub** [reply]
This is really helpful. I'll explore what guardrails I can put around to prevent such issues from happening and decide accordingly.
Thanks a lot @acelaya ! :)
